src/llm/http_llm_calculator.cc

```diff
@@ -580,11 +604,25 @@ class HttpLLMCalculator : public CalculatorBase {
                     } else {  // finish generation
                         OVMS_PROFILE_SCOPE("Generation of last streaming response");
                         std::string response = packIntoServerSideEventMessage(serializeStreamingChunk(this->streamer->end(), finishReason, this->request->getEndpoint()));
-                        response += packIntoServerSideEventMessage("[DONE]");
-                        SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Generated complete streaming response: {}", response);
-                        // Produce last message, but do not produce loopback packets anymore so this is last Process() call
+
+                        // If usage statistics are not supposed to be included, then this is the last chunk
+                        if (!this->request->getStreamOptions().includeUsage)
+                            response += packIntoServerSideEventMessage("[DONE]");
+
                         cc->Outputs().Tag(OUTPUT_TAG_NAME).Add(new OutputDataType{std::move(response)}, timestamp);
+                        SPDLOG_LOGGER_DEBUG(llm_calculator_logger, "Generated complete streaming response: {}", response);
+
+                        // If usage statisctics are included, we trigger next iteration to send additional chunk with usage information
+                        if (this->request->getStreamOptions().includeUsage)
```

Comment:
Why do we need to do it in next iteration? Cant we do it the same as we did with [DONE]? In the same iteration, but pushing multiple events in single MP packet.

Reply:
I have abstracted all that chunk logic (related to serving LLMs) on calculator level so that net_http server does not need to know anything about SSE/OpenAI. It simply takes std::string and sends it to the client.

It's up to calculator to encode the openai format + SSE protocol into std::string. It does not matter if it is 1 or more chunks. Server will send everything you prepare to the client.

Look up the [DONE] chunk, it is done in the same iteration as last text chunk (or it was back when I made the first version of this calculator)